### PR TITLE
Generic udp nodes

### PR DESCRIPTION
**Pull request by flynneva**
- migrate the provided `udp_receiver_node` and `udp_sender_node` to use the `udp_msgs` by default

this should cover *most* use-cases for this package honestly, allowing people to just download and use it right away without modifications. ive tested it locally and it works as expected with one minor quirk: the `udp_receiver_node` does not publish topics if udp packets are streaming in before it is launched. I had to turn off the sensor, then launch the node, then turn on the sensor to be able to see the topics. without this step no `/udp_read` topic was even listed

### CODE REVIEW

**Comment by flynneva** [comment]
still need to fix the unit tests though to handle the new data type (std::vectors rather than just ints)

**Comment by flynneva** [comment]
@JWhitleyWork i cant tell if the problem is in the `io_context/msg_converter` i setup for the `UdpPacket` msg type or if its in the unit tests for the `udp_receiver` and `udp_sender` nodes. i added some prints in the unit test to help try and debug, going offline for tonight but maybe someone else can chime in for some advice.

from the prints it looks like im sending data with the correct size, just i think im probably messing up which data im sending (probably a pointer to the wrong place/missing pointer somewhere)

**Comment by JWhitleyWork** [comment]
@flynneva Would you mind rebasing this on the new `main` (post- #39)? I'd be happy to dig through it after that.

**Comment by flynneva** [comment]
@JWhitleyWork just rebased, should be ready for review.

**Comment by flynneva** [comment]
@JWhitleyWork should be good now

**Comment by flynneva** [comment]
so i tested this today on an actual udp device and the `udp_receiver_node` seemed to be publishing just fine to the `/udp_read` topic as expected. the data looked consistent when I echo'd it out to the console. going to take a second look at the test to see what is going on there.

**Comment by flynneva** [comment]
with my most recent tests it looks like the tests are failing because the `MutBuffer` is pointing to where the data USED to be. sometimes during the test the loop performs fine and actually reads out the correct values - I think this is because randomly that part of the memory hasnt been written over yet so it still contains the old data.

```
4: count is 9, set to true!
4: packet received
4: packet size: 2
4: sum: 41840
4: packet received
4: packet size: 4
4: sum: 41840
4: sum: 13652
4: packet received
4: packet size: 6
4: sum: 0                        <<<< these are what we should expect!
4: sum: 1
4: sum: 3
4: packet received
4: packet size: 8
4: sum: 24000
4: sum: 55394
4: sum: 11841
4: sum: 11841
4: packet received
4: packet size: 10
4: sum: 41840
4: sum: 13652
4: sum: 46419
4: sum: 46419
4: sum: 20499
4: packet received
4: packet size: 12
4: sum: 41840
4: sum: 13652
4: sum: 46419
4: sum: 46419
4: sum: 20499
4: sum: 57847
```

**Comment by flynneva** [comment]
that also should explain why it seems to work when running live on a udp sensor and not on the tests.....

**Comment by JWhitleyWork** [comment]
> that also should explain why it seems to work when running live on a udp sensor and not on the tests.....

This is really starting to look like a threading/race condition issue. Can you try passing 1 to the IoContext instance in the tests and see if the issue persists?

**Comment by flynneva** [comment]
do you just mean like this right at the beginning of the tests?
```
TEST(UdpSenderNodeTest, RosMessageToRawUdpMessageSharedContext)
{
  rclcpp::init(0, nullptr);
  IoContext ctx{1};
```

ran it again after building and still seeing the same issue...works sometimes but other times no:

```
4: count is 9, set to true!
4: packet received
4: packet size: 2
4: sum: 12800
4: packet received
4: packet size: 4
4: sum: 0                                 <<<<  this is right
4: sum: 1
4: packet received
4: packet size: 6
4: sum: 31536
4: sum: 31577
4: sum: 53461
4: packet received
4: packet size: 8
4: sum: 12800
4: sum: 7394
4: sum: 40160
4: sum: 40160
4: packet received
4: packet size: 10
4: sum: 18640
4: sum: 18694
4: sum: 40578
4: sum: 40578
4: sum: 28306
4: packet received
4: packet size: 12
4: sum: 12800
4: sum: 7394
4: sum: 40160
```

**Comment by flynneva** [comment]
i honestly think its the way we structure the test though. we first send the packets, THEN start the receiver. I think we should start the receiver, THEN send the packets, no?

Just i dont know how to start the receiver in the background (probably needs another thread?)
``` 
ASSERT_EQ(
      rclcpp::spin_until_future_complete(
        minimal_publisher, future_2,
        std::chrono::milliseconds(4000)),
      rclcpp::FutureReturnCode::SUCCESS);
    EXPECT_EQ(future_2.get(), true);
  }

  ASSERT_EQ(
    rclcpp::spin_until_future_complete(
      node,
      future_1,
      std::chrono::milliseconds(4000)),
    rclcpp::FutureReturnCode::SUCCESS);
  EXPECT_EQ(future_1.get(), true);
  EXPECT_EQ(sum, 45);
```

**Comment by flynneva** [comment]
you can see this in the test logs too....it sends each packet THEN starts to receive the packets.  I would expect it probably should send a packet & then receive it immediately.

```
4: 0        <<< sent  first packet...should receive this one right after?
4: 1                                 
4: 2
4: 3
4: 4
4: 5
4: 6
4: 7
4: 8
4: 9
4: count is 9, set to true!
4: packet received
4: packet size: 2                    <<<< received first packet...
4: sum: 12800
4: packet received
4: packet size: 4
4: sum: 0
4: sum: 1
4: packet received
4: packet size: 6
4: sum: 31536
4: sum: 31577
```

**Comment by flynneva** [comment]
@JWhitleyWork i think im close. there is no race condition now in the final test and it looks like there is just one more bug to fix in the asyncSend logic somewhere.

**Comment by flynneva** [comment]
@JWhitleyWork all tests passing.

there was some major changes needed to move from `MutBuffer` to `std::vector` in order to get around the race-conditions that we were seeing. not sure what other performance hits we get because of this change.

**Comment by flynneva** [comment]
just let me know if you want me to squash these commits down to one

**Comment by flynneva** [comment]
@reza-ebrahimi also wouldn't mind your review too

File: io_context/src/msg_converters/udp_msgs.cpp [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+{
+  out = in->data;
+  // MutBuffer(const_cast<std::vector<uint8_t> *>(&in->data), sizeof(in->data));
+}
+
+void to_msg(const std::vector<uint8_t> & in, udp_msgs::msg::UdpPacket & out)
+{
+  out.data.resize(in.size());
+  out.data = in;
+  // out.data = *asio::buffer_cast<std::vector<uint8_t> *>(in);
```
Comment (JWhitleyWork) [inline io_context/src/msg_converters/udp_msgs.cpp:40]: Go ahead and remove the commented-out lines.

**Review by JWhitleyWork — CHANGES_REQUESTED** [review]
Just one last item. After that, I'm good to merge. We can profile performance later.

Additionally, would you mind creating a new issue to completely replace `MutBuffer` with `std::vector<uint8_t>`? There's no reason that I know of to use the Boost version instead.

**Comment by flynneva** [comment]
@JWhitleyWork thanks for the review. should be good now and I also cleaned up the commits

**Comment by flynneva** [comment]
last force-push was just to re-word a commit message

**Review by JWhitleyWork — APPROVED** [review]

**Pull request merged**